File: c#/HowMuch/HowMuch.cs
using System.Collections.Generic;
using System.Linq;

namespace DejaVu.HowMuch
{
    public class HowMuch
    {
        public string IsThat(long number)
        {
            if (number == 0) return "zero";

            var digitGroups = DigitGroup.SplitIntoDigitGroups(number);
            var digitGroupsInWords = ProcessDigitGroups(digitGroups);
            return AssembleDigitGroups(digitGroupsInWords);
        }

        private List<string> ProcessDigitGroups(List<DigitGroup> digitGroup)
        {
            var multiples = new[] { "", "thousand", "million", "billion", "trillion", "quadrillion", "quintillion" };

            return digitGroup.Select((dGroup, index) =>
            {
                var digitGroupAsWords = dGroup.AsWords;
                if (string.IsNullOrWhiteSpace(digitGroupAsWords)) return string.Empty;
                return digitGroupAsWords + " " + multiples[digitGroup.Count - (index + 1)];
            }).ToList();
        }

        public string AssembleDigitGroups(List<string> digitGroupsInWords)
        {
            digitGroupsInWords = digitGroupsInWords.Where(r => !string.IsNullOrWhiteSpace(r)).ToList();
            if (digitGroupsInWords.Count == 1)
                return string.Join("", digitGroupsInWords).TrimEnd();

            if (digitGroupsInWords.Last().Contains("and"))
                return string.Join(", ", digitGroupsInWords).TrimEnd();

            for (var i = 0; i < digitGroupsInWords.Count; i++)
            {
                digitGroupsInWords[i] = i < digitGroupsInWords.Count - 2 ? digitGroupsInWords[i] + ", " : digitGroupsInWords[i];
            }

            digitGroupsInWords.Insert(digitGroupsInWords.Count - 1, " and ");

            return string.Join("", digitGroupsInWords).TrimEnd();
        }

    }
}

File: ts_js/HowMuch/HowMuch.ts
class HowMuch {

    public isThat(number: number): string {
        if (number === 0) return "zero";

        var digitGroups = DigitGroup.splitIntoDigitGroups(number);
        var digitGroupsInWords = this.processDigitGroups(digitGroups);
        return this.assembleDigitGroups(digitGroupsInWords);
    }

    private processDigitGroups(digitGroup: DigitGroup[]): string[] {
        var multiples = ["", "thousand", "million", "billion", "trillion", "quadrillion", "quintillion"];

        var result: string[] = [];

        for (var i = 0; i < digitGroup.length; i++) {
            var digitGroupAsWords = digitGroup[i].asWords;
            if (this.isNullOrWhiteSpace(digitGroupAsWords)) {
                result.push("");
                continue;
            }
            result.push(digitGroupAsWords + " " + multiples[digitGroup.length - (i + 1)]);
        }
        return result;
    }

    private isNullOrWhiteSpace(str: string): boolean {
        return str === null || str.match(/^ *$/) !== null;
    }

    public assembleDigitGroups(digitGroupsInWords: string[]): string {
        digitGroupsInWords = digitGroupsInWords.filter(e => { return !this.isNullOrWhiteSpace(e); });

        if (digitGroupsInWords.length === 1)
            return digitGroupsInWords.join("").trim();

        if (digitGroupsInWords[digitGroupsInWords.length - 1].indexOf("and") > -1)
            return digitGroupsInWords.join(", ").trim();

        for (var i = 0; i < digitGroupsInWords.length; i++) {
            digitGroupsInWords[i] = i < digitGroupsInWords.length - 2 ? digitGroupsInWords[i] + ", " : digitGroupsInWords[i];
        }

        digitGroupsInWords.splice(digitGroupsInWords.length - 1, 0, " and ");

        return digitGroupsInWords.join("").trim();
    }
}

